Change extern types to static consts

cdvonbargen opened the pull request:
When building RDKit as a static library, Ricardo found that the externs aren't being resolved correctly causing a number of bad std::any casts. This is his patch (which we apply in the context of the sketcher build) which moves the definitions up from the types.cpp file, and make them static.

cdvonbargen committed:
When building RDKit as a static library, Ricardo found that the externs aren't
being resolved correctly causing a number of bad std::any casts. This is his
patch (which we apply in the context of the sketcher build) which moves the
definitions up from the types.cpp file, and make them static.

d-b-w commented on Code/RDGeneral/types.h:
```diff
@@ -47,232 +47,143 @@ namespace RDKit {
-RDKIT_RDGENERAL_EXPORT extern const std::string
-    _TriposAtomType;  // string Mol2FileParser
-// missing defs for _TriposAtomName//_TriposPartialCharge...
+static const std::string TWOD = "2D";
```
I think you want:

```
static const inline std::string TWOD...
```

for the reasons that Brian mentioned. Maybe constexpr?

d-b-w replied:
https://en.cppreference.com/w/cpp/language/inline.html

cdvonbargen committed:
Use inline keyword to avoid duplication overhead; this still allows the
use of the static keyword to resolve the issue with static libs
https://en.cppreference.com/w/cpp/language/inline.html

cdvonbargen committed:
static inline doesn't work in my static builds; inline only does though

ricrogz commented:
> One solution is to not use std::string but to use const char * so they get placed in the text area and don't need to be initialized.
> 
> For a full optimization, we might need to add const char * key lookup methods to RDProps

Also, `const char*` can be made `constexpr` while `std::string` can't.

greglandrum commented:
> When building RDKit as a static library, Ricardo found that the externs aren't being resolved correctly causing a number of bad std::any casts.

Apologies for what may be a very naive question, but do we have any idea why this is a problem? It's not clear to me why it should be.

Could it be that some dependency isn't being resolved properly and RDGeneral isn't being linked against correctly in either the wasm builds or the static Windows builds that @bp-kelley tried?

bp-kelley commented:
@greglandrum you don't actually have to make the std::string, just reverse the if statement, std::string == has an overload for const char *

```
std::string("prop") == "prop"
```

bp-kelley commented:
That might work, you mean something like

```
constexpr std::string_view PropName= "PropName";
```

It looks like it is just the const char * plus a length so it isn't dynamically allocated.   This has my vote.

bp-kelley commented:
The inline keyword for variables is pretty neat.  I'll have to read up more about them.  No more extern?  Don't mind if I do...

cdvonbargen committed:
inline constexpr std::string_view

cdvonbargen commented:
Update to `an inline constexpr std::string_view defined in a header file`

ricrogz committed:
propagate string_view after static types

ricrogz committed:
add string_view to python converter

ricrogz commented:
@cdvonbargen This is going to require some more changes, starting with https://github.com/rdkit/rdkit/pull/8844. Don't worry, I'm on it.

ricrogz commented:
Almost there, @cdvonbargen. Can you merge master into your branch, so that I can make a PR against it with the final set of changes required for this to work ?

cdvonbargen committed:
casting

greglandrum reviewed (approved):
LGTM

d-b-w commented on Code/RDGeneral/types.h:
```diff
@@ -47,232 +48,144 @@ namespace RDKit {
-RDKIT_RDGENERAL_EXPORT extern const std::string
-    _TriposAtomType;  // string Mol2FileParser
-// missing defs for _TriposAtomName//_TriposPartialCharge...
+constexpr inline std::string_view TWOD = "2D";
```
they mean the same thing... But I see this way, way more often as:

    inline constexpr

on github, the ratio is about 4:1. In schrodinger, it's about 100:1.

https://github.com/search?q=%22inline+constexpr%22&type=code
https://github.com/search?q=%22constexpr+inline%22&type=code

d-b-w reviewed (approved):
really trivial nitpick -  But we should probably do it bc these are the first ones in rdkit, they are what people will copypasta

cdvonbargen committed:
inline-constexpr

greglandrum reviewed (approved):
LGTM

greglandrum committed:
Merge branch 'master' into static_types

greglandrum commented:
@ricrogz @cdvonbargen I just fixed the newly-created merge conflict with this and will merge it once the CI build go through

mbanck-cd commented:
It looks this has changed the API/ABI, e.g. Coot built against librdkit now has a missing symbol, see [Debian Bug #1138824](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1138824):
```
 69s /usr/libexec/coot-1: symbol lookup error: 
/usr/lib/x86_64-linux-gnu/coot/libcoot-lidia-core.so.0: undefined 
symbol: _ZN5RDKit17common_properties8_CIPRankB5cxx11E
```
The demangled symbol is `RDKit::common_properties::_CIPRank`.

In fact, there are only a few `common_properties` symbols left in the shared libraries:
```
mba@curie:/tmp$ for i in 2025.03.1/usr/lib/libRDKit*.so.1.2025.03.1; do nm -D $i | grep common_properties; done | wc -l
345
mba@curie:/tmp$ for i in 2026.03.2/usr/lib/libRDKit*.so.1.2026.03.2; do nm -D $i | grep common_properties; done | wc -l
8
```

greglandrum commented:
> It looks this has changed the API/ABI, e.g. Coot built against librdkit now has a missing symbol, see [Debian Bug #1138824](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1138824):
> 
> ```
> 
>  69s /usr/libexec/coot-1: symbol lookup error: 
> 
> /usr/lib/x86_64-linux-gnu/coot/libcoot-lidia-core.so.0: undefined 
> 
> symbol: _ZN5RDKit17common_properties8_CIPRankB5cxx11E
> 
> ```
> 
> The demangled symbol is `RDKit::common_properties::_CIPRank`.
> 
> 
> 
> In fact, there are only a few `common_properties` symbols left in the shared libraries:
> 
> ```
> 
> mba@curie:/tmp$ for i in 2025.03.1/usr/lib/libRDKit*.so.1.2025.03.1; do nm -D $i | grep common_properties; done | wc -l
> 
> 345
> 
> mba@curie:/tmp$ for i in 2026.03.2/usr/lib/libRDKit*.so.1.2026.03.2; do nm -D $i | grep common_properties; done | wc -l
> 
> 8
> 
> ```

@mbanck-cd we do not guarantee ABI stability from one release to the next. One should assume that dependencies that link against RDKit shared libs will need to be rebuilt against a new RDKit release. Code changes should, however, generally not be required. Any time that is the case, it will be covered in the release notes

mbanck-cd commented:
> @mbanck-cd we do not guarantee ABI stability from one release to the next. One should assume that dependencies that link against RDKit shared libs will need to be rebuilt against a new RDKit release. Code changes should, however, generally not be required. Any time that is the case, it will be covered in the release notes

That's not how shared libraries work, however. You can't change the ABI and then leave [`RDKit_ABI`](https://github.com/rdkit/rdkit/blob/master/CMakeLists.txt#L168) unchanged.

But if that is the policy I guess that means it does not make sense to use RDKit as shared library in Linux distributions and statically linking to it is likely preferred. "Luckily" for everybody(?) RDKit does not seem to be super-popular as shared library and Coot is the only user in Debian/Ubuntu.

greglandrum commented:
> > @mbanck-cd we do not guarantee ABI stability from one release to the next. One should assume that dependencies that link against RDKit shared libs will need to be rebuilt against a new RDKit release. Code changes should, however, generally not be required. Any time that is the case, it will be covered in the release notes
> 
> That's not how shared libraries work, however. You can't change the ABI and then leave [`RDKit_ABI`](https://github.com/rdkit/rdkit/blob/master/CMakeLists.txt#L168) unchanged.
> 
> But if that is the policy I guess that means it does not make sense to use RDKit as shared library in Linux distributions and statically linking to it is likely preferred. "Luckily" for everybody(?) RDKit does not seem to be super-popular as shared library and Coot is the only user in Debian/Ubuntu.

Feel free to handle the packaging of the RDKit in a way that works best for you. I deeply appreciate what you and all the other people who maintain distributions do, but we will continue to do our development in the way that allows us, with our limited resources, to provide a useful and continuously evolving tool for the scientific community.
